core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -154,4 +299,86 @@ public void copyObjectToServer(InetSocketAddress host, KernelOID oid, KernelObje
                     MicroServiceReplicaNotFoundException {
         getServer(host).copyKernelObject(oid, object);
     }
+
+    /**
+     * Method to measure the kernel server metrics with respect to other remote servers and store
+     * them in {@link amino.run.kernel.client.KernelClient#metrics}. These collected metrics are
+     * reported to OMS in heartBeats
+     */
+    private void measureMetrics(final KernelServerInfo serverInfo) {
+        KernelServer server = serverInfo.remoteRef;
+        int dataLength = serverInfo.data.len;
+        NodeMetric metric = serverInfo.metric;
+        int period = serverInfo.metricPollPeriod;
+        serverInfo.refCount++;
+        if (serverInfo.refCount == Integer.MAX_VALUE) {
+            /* Reset to 1 when it is about to overflow */
+            serverInfo.refCountOnHeartBeat = serverInfo.refCount = 1;
+        }
+
+        try {
+            /* Forward time = Request Serialization + Network Delay + Deserialization + Object/call Dispatch time
+            + processing time + Optionally connection establishment time */
+            /* Return time = Response Serialization + Network Delay + Deserialization + Get current nano time at end */
+            /* Forward time would be comparatively higher than return time. But to avoid the clock sync issues
+            between client and server, took the round trip time(RTT) as latency
+             */
+
+            /* Make an empty heartbeat to ensure session is established and cached before measurement */
+            server.receiveHeartBeat();
+
+            /* Measure latency */
+            long t1 = System.nanoTime();
+            server.receiveHeartBeat();
+            long t2 = System.nanoTime();
+
+            /* Measure data transfer rate by sending some arbitrary data */
+            server.receiveHeartBeat(serverInfo.data);
+            long t3 = System.nanoTime();
+            if (t3 - t2 < t2 - t1) {
+                period = MIN_METRIC_POLL_PERIOD_MS;
```

Comment:
With the current values and logic, this is not going to work very well.

For example, for a 10GBit ethernet, with say 50ms latency (a fairly common link between data centers), you will need to send around about 1 GByte of data (which will take about 1 second) for the round trip latency to be about 10% of the total transfer time (so that the measurement is off by less than about 10%).  

Because you are only starting with 1Kbyte, and incrementing by 1KByte every 1 second, it will take 1 million seconds  to get up to 1 Gbyte.  That's about 12 days . And in the process you will waste many, many, many GBytes of transferred data getting there.

I would suggest that you come up with a better plan.  One suggestion might be to double the data size every time (and also perhaps start with a bigger number than 1K).

Reply:
PS I think slow 3G links are a few hundred Kbit/s to 1Mbit/sec, and latencies around 100ms.

https://www.ofcom.org.uk/about-ofcom/latest/media/media-releases/2014/3g-4g-bb-speeds

So the smallest amount of data you're ever likely to need to send (using similar assumptions to above) is around 128KBytes.  You may as well start just below that (say 32Kbytes) and keep doubling it until your data transmission time is significantly greater than your basic line latency (10x in the above example).

In that case, to get from 2^7K (128K) to 2^20K (1G) would take 20-7 = 13 attempts.  That seems fine.  Way better than a million above.

Reply:
Agreed. Doubling the amount of data until your data transmission time is greater than line latency.